src/core/term.rs

```diff
@@ -139,4 +166,70 @@ impl<Witness, App: Application> UntypedProgram<Witness, App> {
     pub fn iter(&self) -> impl Iterator<Item = &Term<Witness, App>> {
         self.0.iter()
     }
+
+    /// Visit the DAG that underlies the program
+    /// and return a stack of node indices that can be popped such that the nodes are in post order.
+    fn get_post_order_stack(&self) -> Vec<usize> {
+        if self.0.is_empty() {
+            return Vec::new();
+        }
+
+        let mut visited = Vec::with_capacity(self.len());
+        let mut to_visit = vec![self.0.len() - 1];
+
+        while let Some(index) = to_visit.pop() {
+            visited.push(index);
+
+            if let Some(l) = self.0[index].get_left() {
+                let idx_l = index - l;
+                if !Self::is_cross_reference(idx_l, &to_visit) {
+                    to_visit.push(idx_l);
+                }
+            }
+            if let Some(r) = self.0[index].get_right() {
+                let idx_r = index - r;
+                if !Self::is_cross_reference(idx_r, &to_visit) {
+                    to_visit.push(idx_r);
+                }
+            }
+        }
+
+        visited
+    }
+
+    /// Return whether the given `index` is a cross-reference to a node that appears
+    /// earlier in the post-order of the DAG.
+    ///
+    /// Using the two-stack approach of [`Self::get_post_order_stack`],
+    /// the largest index of any unvisited node is always the last element of `to_visit`.
+    /// In a post order, the indices of descendants are always lower than the ancestor index.
+    fn is_cross_reference(index: usize, to_visit: &[usize]) -> bool {
+        if let Some(largest_unvisited_index) = to_visit.last() {
+            if index <= *largest_unvisited_index {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Return whether the program is in _canonical order_.
+    /// This means that node indices appear in post order,
+    /// i.e., children appear before their parent.
+    /// It also means that if there is a node that cannot be reached from the root,
+    /// then the program is not in canonical order.
```

Comment:
In 1cef378b9dcf712fa46853b9d96d55e4c865a70a:

This comment is at least partially incorrect (it's not about children appearing before their parent, which is always trivially true, but about left children appearing before right children). We should also clarify that we **also** check that there are no disconnected nodes, in the last line of the function. As written it looks like the ordering check somehow can detect disconnected nodes by itself.

Curiously the C code seems to have an assertion that `bottom == self.len` and no comment about checking for disconnected nodes ... can you trigger this assertion in the C code?

Reply:
I updated the wording to clarify.